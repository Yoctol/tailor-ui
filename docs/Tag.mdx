---
name: Tag
menu: Components
route: /components/tag
---

import { PropsTable, Playground } from 'docz';
import { Input, Button, Flex, Box, Tag } from 'tailor-ui';

# Tag

Tag for categorizing or markup.

## When To Use

- It can be used to tag by dimension or property.
- When categorizing.

## Examples

```js
import { Tag } from 'tailor-ui';
```

### Basic

<Playground>
  <Tag>Content</Tag>
  <Tag>Content</Tag>
</Playground>

### Closable

<Playground>
  <Tag closable>Content</Tag>
  <Tag closable>Content</Tag>
  <Tag closable>Content</Tag>
  <Tag closable>Content</Tag>
</Playground>

### Add & Remove Dynamically

<Playground>
  {() => {
    const [values, setValues] = React.useState(['tag 1', 'tag 2', 'tag 3']);
    const [editing, setEditing] = React.useState(false);

    const handleCreateTag = event => {
      const value = event.target.value.trim();

      if (value !== '') {
        setValues([...values, value]);
      }

      setEditing(false);
    };

    return (
      <Flex>
        {values.map((value, index) => (
          <Tag
            key={value}
            closable={index !== 0}
            onClosed={() => setValues(values.filter(v => v !== value))}
          >
            {value}
          </Tag>
        ))}
        <Flex ml="2">
          {editing ? (
            <Input
              width="80px"
              size="sm"
              defaultValue=""
              autoFocus
              onBlur={handleCreateTag}
              onPressEnter={handleCreateTag}
            />
          ) : (
            <Button text size="sm" onClick={() => setEditing(true)}>
              + Add Tag
            </Button>
          )}
        </Flex>
      </Flex>
    );

}}

</Playground>

## API

<!-- FIXME: API Table -->
<!-- <PropsTable of={Tag} /> -->
