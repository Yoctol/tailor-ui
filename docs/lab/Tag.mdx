---
name: Tag
menu: Lab
route: /lab/tag
---

import { PropsTable, Playground } from 'docz';
import { Input, Button, Flex, Box, useModal } from 'tailor-ui';
import { Tag } from '@tailor-ui/lab';

# Tag

Tag for categorizing or markup.

## When To Use

- It can be used to tag by dimension or property.
- When categorizing.

## Examples

```js
import { Tag } from '@tailor-ui/lab';
```

### Basic

<Playground>
  <Tag>Content</Tag>
  <Tag invalid>Content</Tag>
</Playground>

### Use prefix

<Playground>
  <Tag prefix="10">Content</Tag>
  <Tag prefix="20" invalid closable>
    Content
  </Tag>
</Playground>

### Editable

<Playground>
  <Tag prefix="10" editable onChange={console.log}>
    Content
  </Tag>
  <Tag prefix="20" editable closable onChange={console.log}>
    Content
  </Tag>
</Playground>

### Closable

<Playground>
  <Tag closable>Content</Tag>
  <Tag closable>Content</Tag>
  <Tag closable>Content</Tag>
  <Tag closable>Content</Tag>
</Playground>

### Use with canClose

<Playground>
  {() => {
    const modal = useModal();

    return (
      <Tag
        closable
        canClose={() => modal.error({
          cancelable: true,
          title: 'Delete this tag?',
          content: 'Are you sure? it can not be rollback!',
        })}
      >
        DELETE ME
      </Tag>
    );

}}

</Playground>

### Add & Remove Dynamically

<Playground>
  {() => {
    const [values, setValues] = React.useState(['tag 1', 'tag 2', 'tag 3']);
    const [editing, setEditing] = React.useState(false);

    return (
      <Flex flexWrap="wrap">
        {values.map((value, index) => {
          const invalid = values.filter((_, i) => i !== index).includes(value);

          return (
            <Flex key={value} flexDirection="column">
              <Tag
                key={value}
                closable
                editable
                invalid={invalid}
                onClosed={() => setValues(values.filter(v => v !== value))}
                onChange={(prevValue, value) => {
                  if (value.trim() !== '') {
                    setValues(prevValues =>
                      prevValues.map((originalValue, id) =>
                        originalValue === prevValue && id === index ? value : originalValue
                      )
                    );
                    setEditing(false);
                  }
                }}
              >
                {value}
              </Tag>
              {invalid && <Box mt="1" mr="1" color="danger" fontSize="sm">duplicate tag!</Box>}
            </Flex>
          );
        })}
        {values.length === new Set(values).size && (
          editing ? (
            <Tag
              editable
              initialEditing
              onBlur={() => setEditing(false)}
              onChange={(_ , value) => {
                if (value.trim() !== '') {
                  setValues(prevValues => ([...prevValues, value]));
                }
                setEditing(false);
              }}
            />
          ) : (
            <Button
              size="sm"
              variant="regular"
              onClick={() => setEditing(true)}
            >
              + Add Tag
            </Button>
          )
        )}
      </Flex>
    );

}}

</Playground>

## API

<PropsTable of={Tag} />
