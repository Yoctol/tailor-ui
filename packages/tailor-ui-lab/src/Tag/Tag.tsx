import React, {
  FocusEvent,
  FunctionComponent,
  KeyboardEvent,
  MouseEventHandler,
  ReactNode,
  useState,
} from 'react';
import styled from 'styled-components';
import { MdClose } from 'react-icons/md';
import { animated, useSpring } from 'react-spring';

import { Icon, useMeasure, usePrevious } from 'tailor-ui';

import { StyledTag, StyledTagInput, StyledTagPrefix } from './styles';

const StyledTagWrapper = styled.div`
  display: inline-flex;
  overflow-x: hidden;
`;

const AnimatedStyledTagWrapper = animated(StyledTagWrapper);

const StyledCloseIcon = styled(Icon)`
  & svg {
    fill: ${p => p.theme.colors.primaryLight2};
  }

  &:hover svg {
    fill: ${p => p.theme.colors.primaryLight};
  }
`;

const CloseIcon: FunctionComponent<{
  onClick: MouseEventHandler;
  invalid: boolean;
}> = ({ onClick, invalid }) => (
  <StyledCloseIcon
    size="16"
    ml="1"
    cursor="pointer"
    fill={invalid ? 'danger' : undefined}
    onClick={onClick}
    type={MdClose}
  />
);

export interface TagProps {
  /**
   * Whether the Tag can be closed
   */
  closable?: boolean;
  editable?: boolean;
  /**
   * Callback executed when close animation is completed
   */
  onClosed?: () => void;
  canClose?: () => boolean;
  onChange?: (previousValue: string, value: string) => void;
  children?: string;
  initialEditing?: boolean;
  invalid?: boolean;
  prefix?: ReactNode;
}

const Tag: FunctionComponent<TagProps> = ({
  children = '',
  editable = false,
  initialEditing = false,
  closable,
  invalid = false,
  onClosed,
  canClose,
  prefix,
  onChange,
  ...otherProps
}) => {
  const [on, setOn] = useState(true);
  const [editing, setEditing] = useState(initialEditing);
  const previous = usePrevious(on);
  const [bind, { width }] = useMeasure();

  const props = useSpring({
    opacity: on ? 1 : 0,
    marginRight: on ? 4 : 0,
    width: on ? width : 0,
    onRest: ({ width: restWidth }: { width: number }) => {
      if (!on && restWidth === 0 && onClosed) {
        onClosed();
      }
    },
  });

  const handleUpdate = (
    event: FocusEvent<HTMLInputElement> | KeyboardEvent<HTMLInputElement>
  ) => {
    const { value } = event.currentTarget;

    setEditing(false);

    if (onChange && children !== value) {
      onChange(children, value);
    }
  };

  return (
    <AnimatedStyledTagWrapper
      style={{ ...props, width: on && previous === on ? 'auto' : props.width }}
      {...bind}
    >
      <StyledTag
        editable={editable}
        invalid={invalid}
        onClick={() => {
          if (editable) {
            setEditing(true);
          }
        }}
        {...otherProps}
      >
        {prefix && (
          <StyledTagPrefix invalid={invalid}>{prefix}</StyledTagPrefix>
        )}
        {editing ? (
          <StyledTagInput
            autoFocus
            invalid={invalid}
            defaultValue={children}
            onBlur={handleUpdate}
            onKeyPress={event => {
              if (event.key === 'Enter') {
                handleUpdate(event);
              }
            }}
            onKeyDown={event => {
              if (event.key === 'Escape') {
                setEditing(false);
              }
            }}
          />
        ) : (
          children
        )}
        {closable && (
          <CloseIcon
            invalid={invalid}
            onClick={async event => {
              event.stopPropagation();

              if (canClose && (await !canClose())) {
                return;
              }

              setOn(!on);
            }}
          />
        )}
      </StyledTag>
    </AnimatedStyledTagWrapper>
  );
};

export default Tag;
