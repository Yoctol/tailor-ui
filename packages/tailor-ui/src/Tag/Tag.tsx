import React, { FunctionComponent, PureComponent, useState } from 'react';
import styled from 'styled-components';
import { MdClose } from 'react-icons/md';
import { animated, config, useSpring } from 'react-spring';
import { omit } from 'ramda';

import Icon from '../Icon';
import tag from '../utils/CleanTag';

const StyledTag = styled(tag.div)`
  display: inline-flex;
  align-items: center;
  height: ${p => p.theme.heights.sm};
  padding: 0 ${p => p.theme.paddings.xs};
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  border-color: ${p => p.theme.colors.gray400};
  font-size: ${p => p.theme.fontSizes.sm};
  white-space: nowrap;
  cursor: pointer;
  ${p => p.theme.transition};

  &:hover {
    border-color: ${p => p.theme.colors.gray700};
  }
`;

const CloseIcon = styled(Icon)`
  &:hover svg {
    fill: ${p => p.theme.colors.gray700};
  }
`;

const StyledTagWrapper = styled.div`
  display: inline-flex;
  height: ${p => p.theme.heights.sm};

  & + & {
    margin-left: 8px;
  }
`;

const AnimatedStyledTagWrapper = animated(StyledTagWrapper);

interface ClosableTagProps {
  onClosed?: () => void;
}

const ClosableTag: FunctionComponent<ClosableTagProps> = ({
  children,
  onClosed,
  ...props
}) => {
  const [on, setOn] = useState(true);

  const styles = useSpring({
    onRest: ({ width }: any) => {
      if (width === 0 && onClosed) {
        onClosed();
      }
    },
    immediate: on,
    opacity: on ? 1 : 0,
    width: on ? 'auto' : 0,
    marginLeft: on ? 8 : 0,
    config: config.stiff,
  });

  return (
    <AnimatedStyledTagWrapper style={styles}>
      <div style={{ display: 'inline-flex' }}>
        <StyledTag {...props}>
          {children}
          <CloseIcon
            size="16"
            ml="1"
            fill="gray400"
            cursor="pointer"
            onClick={() => setOn(!on)}
            type={MdClose}
          />
        </StyledTag>
      </div>
    </AnimatedStyledTagWrapper>
  );
};

const BaseTag: FunctionComponent = ({ children, ...props }) => (
  <StyledTagWrapper>
    <StyledTag {...omit(['onClosed'], props)}>{children}</StyledTag>
  </StyledTagWrapper>
);

export interface TagProps {
  /**
   * Whether the Tag can be closed
   */
  closable?: boolean;
  /**
   * Callback executed when close animation is completed
   */
  onClosed?: () => void;
}

class Tag extends PureComponent<TagProps> {
  static defaultProps = {
    closable: false,
    onClosed: () => {},
  };

  render() {
    const { closable, children, ...props } = this.props;
    const RenderComponent = closable ? ClosableTag : BaseTag;

    return <RenderComponent {...props}>{children}</RenderComponent>;
  }
}

export default Tag;
