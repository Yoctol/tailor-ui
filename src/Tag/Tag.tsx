import React, {
  ComponentPropsWithoutRef,
  FC,
  FocusEvent,
  KeyboardEvent,
  MouseEventHandler,
  ReactNode,
  forwardRef,
  useCallback,
  useState,
} from 'react';
import styled from 'styled-components';
import { MdClose } from 'react-icons/md';
import { animated, useSpring } from 'react-spring';

import { Icon } from '../Icon';
import { useMeasure, usePrevious } from '../hooks';

import { StyledTag, StyledTagInput, StyledTagPrefix } from './styles';

const StyledTagWrapper = styled.div`
  display: inline-flex;
  overflow-x: hidden;
`;

const AnimatedStyledTagWrapper = animated(StyledTagWrapper);

const StyledCloseIcon = styled(Icon)`
  & svg {
    fill: ${(p) => p.theme.colors.primaryLight2};
  }

  &:hover svg {
    fill: ${(p) => p.theme.colors.primaryLight};
  }
`;

interface CloseIconProps {
  onClick: MouseEventHandler;
  invalid: boolean;
  disabled: boolean;
}

const CloseIcon: FC<CloseIconProps> = ({ onClick, disabled, invalid }) => {
  let fill: string | undefined;

  if (invalid) {
    fill = 'danger';
  }
  if (disabled) {
    fill = 'gray500';
  }

  return (
    <StyledCloseIcon
      size="16"
      ml="1"
      cursor={disabled ? 'not-allowed' : 'pointer'}
      fill={fill}
      onClick={onClick}
      type={MdClose}
    />
  );
};

export interface TagProps
  extends Omit<ComponentPropsWithoutRef<'div'>, 'prefix' | 'onChange'> {
  /**
   * Whether the Tag can be closed
   */
  closable?: boolean;
  editable?: boolean;
  /**
   * Callback executed when close animation is completed
   */
  onClosed?: () => void;
  onClose?: () => void;
  canClose?: () => boolean | Promise<boolean>;
  onChange?: (previousValue: string, value: string) => void;
  children?: string;
  initialEditing?: boolean;
  invalid?: boolean;
  disabled?: boolean;
  prefix?: ReactNode;
}

const Tag = forwardRef<HTMLDivElement, TagProps>(function Tag(
  {
    children = '',
    editable = false,
    initialEditing = false,
    closable,
    invalid = false,
    disabled = false,
    onClose,
    onClosed,
    canClose,
    prefix,
    onChange,
    onClick,
    ...otherProps
  },
  ref
) {
  const [on, setOn] = useState(true);
  const [editing, setEditing] = useState(initialEditing);
  const previous = usePrevious(on);
  const [bind, { width }] = useMeasure();

  const props = useSpring({
    opacity: on ? 1 : 0,
    marginRight: on ? 4 : 0,
    width: on ? width : 0,
    onRest: ({ width: restWidth }: { width: number }) => {
      if (!on && restWidth === 0 && onClosed) {
        onClosed();
      }
    },
  });

  const handleUpdate = useCallback(
    (event: FocusEvent<HTMLInputElement> | KeyboardEvent<HTMLInputElement>) => {
      const { value } = event.currentTarget;

      setEditing(false);

      if (onChange && children !== value) {
        onChange(children, value);
      }
    },
    [children, onChange]
  );

  return (
    <AnimatedStyledTagWrapper
      style={{ ...props, width: on && previous === on ? 'auto' : props.width }}
      {...bind}
    >
      <StyledTag
        ref={ref}
        editable={editable}
        disabled={disabled}
        invalid={invalid}
        onClick={(event) => {
          if (disabled) {
            return;
          }

          if (editable) {
            setEditing(true);
          }
          if (onClick) {
            onClick(event);
          }
        }}
        {...otherProps}
      >
        {prefix && (
          <StyledTagPrefix invalid={invalid}>{prefix}</StyledTagPrefix>
        )}
        {editing ? (
          <StyledTagInput
            autoFocus
            defaultValue={children}
            onBlur={handleUpdate}
            onKeyPress={(event) => {
              if (event.key === 'Enter') {
                handleUpdate(event);
              }
            }}
            onKeyDown={(event) => {
              if (event.key === 'Escape') {
                setEditing(false);
              }
            }}
          />
        ) : (
          children
        )}
        {closable && (
          <CloseIcon
            invalid={invalid}
            disabled={disabled}
            onClick={async (event) => {
              if (disabled) {
                return;
              }

              event.stopPropagation();

              if (canClose && !(await canClose())) {
                return;
              }

              if (onClose) {
                onClose();
              } else {
                setOn(false);
              }
            }}
          />
        )}
      </StyledTag>
    </AnimatedStyledTagWrapper>
  );
});

export { Tag };
