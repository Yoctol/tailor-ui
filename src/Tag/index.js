import CloseIcon from 'react-icons/lib/md/close';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import styled from 'styled-components';
import { Spring, animated } from 'react-spring';
import { Toggle } from 'react-powerplug';

import Icon from '../Icon';

const StyledTag = styled.div`
  display: inline-flex;
  align-items: center;
  padding: ${p => p.theme.space.paddingYSm} ${p => p.theme.space.paddingXSm};
  border: ${p => p.theme.borders.default};
  border-radius: ${p => p.theme.radii[2]};
  border-color: ${p => p.theme.colors.gray[7]};
  font-size: ${p => p.theme.fontSizes.sm};
  white-space: nowrap;
  cursor: pointer;
  ${p => p.theme.transition};

  &:hover {
    border-color: ${p => p.theme.colors.gray[3]};
  }
`;

const ClosableIcon = styled(Icon)`
  &:hover svg {
    fill: ${p => p.theme.colors.gray[3]};
  }
`;

const StyledTagWrapper = styled.div`
  display: inline-flex;

  & + & {
    margin-left: 8px;
  }
`;

const AnimatedStyledTagWrapper = animated(StyledTagWrapper);

const AnimatedTag = ({ children, onClosed, ...props }) => (
  <Toggle initial>
    {({ on, toggle }) => (
      <Spring
        native
        onRest={({ width }) => width === 0 && onClosed()}
        from={{ transform: 'scale(1)', opacity: 1, width: 'auto' }}
        to={{
          transform: on ? 'scale(1)' : 'scale(0)',
          opacity: on ? 1 : 0,
          width: on ? 'auto' : 0,
          marginLeft: on ? undefined : 0,
        }}
      >
        {styles => (
          <AnimatedStyledTagWrapper style={styles}>
            <StyledTag {...props}>
              {children}
              <ClosableIcon
                size="16"
                ml="1"
                fill="gray.7"
                cursor="pointer"
                onClick={toggle}
                type={CloseIcon}
              />
            </StyledTag>
          </AnimatedStyledTagWrapper>
        )}
      </Spring>
    )}
  </Toggle>
);

const BaseTag = ({ children, ...props }) => (
  <StyledTagWrapper>
    <StyledTag {...props}>{children}</StyledTag>
  </StyledTagWrapper>
);

class Tag extends PureComponent {
  render() {
    const { closable, children, ...props } = this.props;
    const RenderComponent = closable ? AnimatedTag : BaseTag;

    return <RenderComponent {...props}>{children}</RenderComponent>;
  }
}

Tag.propTypes = {
  /**
   * Whether the Tag can be closed
   */
  closable: PropTypes.bool,
  /**
   * Callback executed when close animation is completed
   */
  onClosed: PropTypes.func,
};

Tag.defaultProps = {
  closable: false,
  onClosed: () => {},
};

export default Tag;
