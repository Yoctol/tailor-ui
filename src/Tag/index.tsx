import React, { FunctionComponent, PureComponent, useState } from 'react';
import { MdClose } from 'react-icons/md';
import { Spring, animated } from 'react-spring';
import { omit } from 'ramda';

import styled from 'utils/styled-components';
import tag from 'utils/CleanTag';

import Icon from '../Icon';

const StyledTag = styled(tag.div)`
  display: inline-flex;
  align-items: center;
  height: ${p => p.theme.heights.sm};
  padding: 0 ${p => p.theme.paddings.xs};
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  border-color: ${p => p.theme.colors.gray400};
  font-size: ${p => p.theme.fontSizes.sm};
  white-space: nowrap;
  cursor: pointer;
  ${p => p.theme.transition};

  &:hover {
    border-color: ${p => p.theme.colors.gray700};
  }
`;

const CloseIcon = styled<any, any>(Icon)`
  &:hover svg {
    fill: ${p => p.theme.colors.gray700};
  }
`;

const StyledTagWrapper = styled.div`
  display: inline-flex;
  height: ${p => p.theme.heights.sm};

  & + & {
    margin-left: 8px;
  }
`;

const AnimatedStyledTagWrapper = animated(StyledTagWrapper);

interface IClosableTagProps {
  onClosed?: () => void;
}

const ClosableTag: FunctionComponent<IClosableTagProps> = ({
  children,
  onClosed,
  ...props
}) => {
  const [on, setOn] = useState(true);

  return (
    <Spring
      native
      onRest={({ width }: any) => {
        if (width === 0 && onClosed) {
          onClosed();
        }
      }}
      from={{ transform: 'scale(1)', opacity: 1, width: 'auto' }}
      to={{
        transform: on ? 'scale(1)' : 'scale(0)',
        opacity: on ? 1 : 0,
        width: on ? 'auto' : 0,
        marginLeft: on ? undefined : 0,
      }}
    >
      {styles => (
        <AnimatedStyledTagWrapper style={styles}>
          <StyledTag {...props}>
            {children}
            <CloseIcon
              size="16"
              ml="1"
              fill="gray400"
              cursor="pointer"
              onClick={() => setOn(!on)}
              type={MdClose}
            />
          </StyledTag>
        </AnimatedStyledTagWrapper>
      )}
    </Spring>
  );
};

const BaseTag: FunctionComponent = ({ children, ...props }) => (
  <StyledTagWrapper>
    <StyledTag {...omit(['onClosed'], props)}>{children}</StyledTag>
  </StyledTagWrapper>
);

export interface ITagProps {
  /**
   * Whether the Tag can be closed
   */
  closable?: boolean;
  /**
   * Callback executed when close animation is completed
   */
  onClosed?: () => void;
}

class Tag extends PureComponent<ITagProps> {
  static defaultProps = {
    closable: false,
    onClosed: () => {},
  };

  render() {
    const { closable, children, ...props } = this.props;
    const RenderComponent = closable ? ClosableTag : BaseTag;

    return <RenderComponent {...props}>{children}</RenderComponent>;
  }
}

export default Tag;
