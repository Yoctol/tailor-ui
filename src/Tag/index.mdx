---
name: Tag
menu: Components
route: /components/tag
---

import { PropsTable, Playground } from 'docz';
import { Set, Toggle } from 'react-powerplug';

import Input from '../Input';
import Button from '../Button';
import Flex from '../Grid/Flex';
import Box from '../Grid/Box';
import Tag from './';

# Tag

Tag for categorizing or markup.

## When To Use

- It can be used to tag by dimension or property.
- When categorizing.

## Examples

```js
import { Tag } from '@yoctol/ui';
```

### Basic

<Playground>
  <Tag>Content</Tag>
  <Tag>Content</Tag>
</Playground>

### Closable

<Playground>
  <Tag closable>Content</Tag>
  <Tag closable>Content</Tag>
  <Tag closable>Content</Tag>
  <Tag closable>Content</Tag>
</Playground>

### Add & Remove Dynamically

<Playground>
  <Set initial={['tag 1', 'tag 2', 'tag 3']}>
    {({ values, remove, add }) => (
      <Toggle>
        {({ on, toggle }) => (
          <Flex>
            {values.map((value, index) => (
              <Tag
                key={value}
                closable={index !== 0}
                onClosed={() => remove(value)}
              >
                {value}
              </Tag>
            ))}
            <Flex ml="2">
              {on ? (
                <Input
                  width="80px"
                  size="sm"
                  defaultValue=""
                  autoFocus
                  onBlur={event => {
                    const value = event.target.value.trim();
                    value !== '' && add(value);
                    toggle();
                  }}
                  onPressEnter={event => {
                    const value = event.target.value.trim();
                    value !== '' && add(value);
                    toggle();
                  }}
                />
              ) : (
                <Button text size="sm" onClick={toggle}>+ Add Tag</Button>
              )}
            </Flex>
          </Flex>
        )}
      </Toggle>
    )}
  </Set>
</Playground>

## API

<PropsTable of={Tag} />
