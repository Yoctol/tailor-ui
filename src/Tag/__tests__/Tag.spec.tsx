import React from 'react';
import userEvent from '@testing-library/user-event';

import { render, screen, waitFor } from 'test/test-utils';

import { Tag } from '../Tag';

describe('Tag', () => {
  it('should render tag', () => {
    const { container } = render(<Tag>Tailor UI</Tag>);

    expect(container).toMatchSnapshot();
  });

  it('should render tag with prefix', () => {
    const { container } = render(<Tag prefix="1">Tailor UI</Tag>);

    expect(container).toMatchSnapshot();
  });

  it('should render invalid style tag', () => {
    const { container } = render(<Tag invalid>Tailor UI</Tag>);

    expect(container).toMatchSnapshot();
  });

  it('should render editable tag and call onChange', async () => {
    const onChange = jest.fn();
    const { container } = render(
      <Tag editable onChange={onChange}>
        Foo
      </Tag>
    );

    const tagText = screen.getByText('Foo');
    userEvent.click(tagText);

    const input = container.querySelector('input') as HTMLInputElement;

    userEvent.type(input, '{selectall}{backspace}Foo Bar');
    expect(input.value).toBe('Foo Bar');

    userEvent.type(input, '{enter}');

    expect(onChange).toBeCalledWith('Foo', 'Foo Bar');
  });

  it('should render closable tag', async () => {
    const { container } = render(<Tag closable>Tag A</Tag>);

    const closeIcon = container.querySelector('i');

    userEvent.click(closeIcon as HTMLElement);
  });

  it('should call onClosed when close icon is clicked', async () => {
    const onClosed = jest.fn();

    const { container } = render(
      <Tag closable onClosed={onClosed}>
        Tag A
      </Tag>
    );

    const closeIcon = container.querySelector('i');

    userEvent.click(closeIcon as HTMLElement);

    await waitFor(() => expect(onClosed).toBeCalled());
  });

  it('should call canClose when close icon is clicked and still visible', () => {
    const canClose = jest.fn().mockResolvedValue(false);

    const { container } = render(
      <Tag closable canClose={canClose}>
        Tailor UI
      </Tag>
    );

    const closeIcon = container.querySelector('i');

    userEvent.click(closeIcon as HTMLElement);

    const tag = screen.getByText('Tailor UI');

    expect(canClose).toBeCalled();
    expect(tag).toBeInTheDocument();
  });

  it('should call canClose when close icon is clicked and close correct', async () => {
    const canClose = jest.fn().mockResolvedValue(true);

    const { container } = render(
      <Tag closable canClose={canClose}>
        Tailor UI
      </Tag>
    );

    const closeIcon = container.querySelector('i');

    userEvent.click(closeIcon as HTMLElement);

    expect(canClose).toBeCalled();
  });
});
