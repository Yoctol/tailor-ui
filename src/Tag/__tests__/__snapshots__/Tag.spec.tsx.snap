// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag should render invalid style tag 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 28px;
  padding: 0 8px;
  border: 1px solid;
  border-radius: 0.5rem;
  border-color: #ff5757;
  background-color: #ffffff;
  color: #ff5757;
  font-size: 0.875rem;
  white-space: nowrap;
  cursor: pointer;
}

.c1 input {
  color: #ff5757;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow-x: hidden;
}

<div>
  <div
    class="c0"
    style="opacity: 1; margin-right: 4px; width: 0px;"
  >
    <div
      class="c1"
    >
      Tailor UI
    </div>
  </div>
</div>
`;

exports[`Tag should render tag 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 28px;
  padding: 0 8px;
  border: 1px solid;
  border-radius: 0.5rem;
  border-color: #3990cd;
  background-color: #fbfbfd;
  color: #3990cd;
  font-size: 0.875rem;
  white-space: nowrap;
  cursor: pointer;
}

.c1 input {
  color: #3990cd;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow-x: hidden;
}

<div>
  <div
    class="c0"
    style="opacity: 1; margin-right: 4px; width: 0px;"
  >
    <div
      class="c1"
    >
      Tailor UI
    </div>
  </div>
</div>
`;

exports[`Tag should render tag with prefix 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 28px;
  padding: 0 8px;
  border: 1px solid;
  border-radius: 0.5rem;
  border-color: #3990cd;
  background-color: #fbfbfd;
  color: #3990cd;
  font-size: 0.875rem;
  white-space: nowrap;
  cursor: pointer;
}

.c1 input {
  color: #3990cd;
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  margin-right: 8px;
  padding-right: 8px;
  border-right: 1px solid;
  border-color: #3990cd;
  color: #3990cd;
  font-size: 0.875rem;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow-x: hidden;
}

<div>
  <div
    class="c0"
    style="opacity: 1; margin-right: 4px; width: 0px;"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        1
      </span>
      Tailor UI
    </div>
  </div>
</div>
`;
